In [3]:
%load_ext sql

In [4]:
%sql postgresql://postgres:root@localhost/sqlda  

In [5]:
pg_version=%sql select version()
print(pg_version)

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.
+------------------------------------------------------------+
|                          version                           |
+------------------------------------------------------------+
| PostgreSQL 12.4, compiled by Visual C++ build 1914, 64-bit |
+------------------------------------------------------------+


In [6]:
%sql \dt

 * postgresql://postgres:***@localhost/sqlda
29 rows affected.


[('public', 'bat_ltd_sales', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_count', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_delay', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_growth', 'table', 'postgres'),
 ('public', 'bat_ltd_sales_vol', 'table', 'postgres'),
 ('public', 'bat_sales', 'table', 'postgres'),
 ('public', 'bat_sales_daily', 'table', 'postgres'),
 ('public', 'bat_sales_daily_delay', 'table', 'postgres'),
 ('public', 'bat_sales_delay_vol', 'table', 'postgres'),
 ('public', 'bat_sales_growth', 'table', 'postgres'),
 ('public', 'closest_dealerships', 'table', 'postgres'),
 ('public', 'countries', 'table', 'postgres'),
 ('public', 'customer_sales', 'table', 'postgres'),
 ('public', 'customer_survey', 'table', 'postgres'),
 ('public', 'customers', 'table', 'postgres'),
 ('public', 'dealerships', 'table', 'postgres'),
 ('public', 'emails', 'table', 'postgres'),
 ('public', 'lemon_sales', 'table', 'postgres'),
 ('public', 'lemon_sales_count', 'table', 'postgres'),
 ('public', 'lemon_sales_delay', 'table', 'postgres'),
 ('public', 'lemon_sales_growth', 'table', 'postgres'),
 ('public', 'lemon_sales_sum', 'table', 'postgres'),
 ('public', 'product_names', 'table', 'postgres'),
 ('public', 'products', 'table', 'postgres'),
 ('public', 'products_sales', 'table', 'postgres'),
 ('public', 'public_transportation_by_zip', 'table', 'postgres'),
 ('public', 'sales', 'table', 'postgres'),
 ('public', 'salespeople', 'table', 'postgres'),
 ('public', 'top_cities_data', 'table', 'postgres')]

#### Analyzing the Difference in the Sales Price Hypothesis
Investigate the hypothesis that the reduction in
sales growth can be attributed to the price point of the Bat Scooter. Previously, we
considered the launch date. However, there could be another factor – the sales price
included. If we consider the product list of scooters, and exclude
the Bat model scooter, we can see that there are two price categories, 699.99 and
above, or 499.99 and below. The Bat Scooter sits exactly between these two groups;
perhaps the reduction in sales growth can be attributed to the different pricing model.
In this activity, we will test this hypothesis by comparing Bat sales to the 2013 Lemon.

##### Select the sales_transaction_date column from the 2013 Lemon sales and insert the column into a table called lemon_sales

In [17]:
%%sql
SELECT customer_id, sales_transaction_date  INTO lemon_sales FROM sales WHERE
product_id=3;

 * postgresql://postgres:***@localhost/sqlda
16558 rows affected.


[]

##### Count the sales records available for the 2013 Lemon

In [18]:
%%sql
SELECT count(sales_transaction_date) AS Lemon_Scooter_2013_Sales FROM lemon_sales;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(16558,)]

##### Latest sales_transaction_date

In [19]:
%%sql
SELECT max(sales_transaction_date) FROM lemon_sales;

 * postgresql://postgres:***@localhost/sqlda
1 rows affected.


[(datetime.datetime(2018, 12, 27, 19, 12, 10),)]

Production between May 2013 and December 2018

##### Convert the sales_transaction_date column to a date type

In [20]:
%%sql
ALTER TABLE lemon_sales ALTER COLUMN sales_transaction_date TYPE
DATE;

 * postgresql://postgres:***@localhost/sqlda
Done.


[]

We are converting the datatype from DATE_TIME to DATE so as to remove the time
information from the field. We are only interested in accumulating numbers,
but just the date and not the time. Hence, it is easier just to remove the time
information from the field.

##### Count the number of sales per day within the lemon_sales table and insert this figure into a table called lemon_sales_count

In [8]:
%%sql
SELECT sales_transaction_date, 
       COUNT(sales_transaction_date) AS sale_per_day 
INTO lemon_sales_count
FROM lemon_sales 
GROUP BY sales_transaction_date
ORDER BY sales_transaction_date;

 * postgresql://postgres:***@localhost/sqlda
2058 rows affected.


[]

In [10]:
%%sql
SELECT * 
FROM lemon_sales_count 
LIMIT 5;

 * postgresql://postgres:***@localhost/sqlda
5 rows affected.


[(datetime.date(2013, 5, 1), 6),
 (datetime.date(2013, 5, 2), 8),
 (datetime.date(2013, 5, 3), 4),
 (datetime.date(2013, 5, 4), 9),
 (datetime.date(2013, 5, 5), 9)]

##### Calculate the cumulative sum of sales and insert the corresponding table into a new table labeled lemon_sales_sum

In [18]:
%%sql
SELECT sales_transaction_date, 
       lemon_sales_count.sale_per_day,
       sum(sale_per_day) OVER (ORDER BY sales_transaction_date)
INTO lemon_sales_sum 
FROM lemon_sales_count
GROUP BY sales_transaction_date, lemon_sales_count.sale_per_day
ORDER BY sales_transaction_date;

 * postgresql://postgres:***@localhost/sqlda
2058 rows affected.


[]

In [19]:
%%sql 
SELECT * FROM lemon_sales_sum LIMIT 5;

 * postgresql://postgres:***@localhost/sqlda
5 rows affected.


[(datetime.date(2013, 5, 1), 6, Decimal('6')),
 (datetime.date(2013, 5, 2), 8, Decimal('14')),
 (datetime.date(2013, 5, 3), 4, Decimal('18')),
 (datetime.date(2013, 5, 4), 9, Decimal('27')),
 (datetime.date(2013, 5, 5), 9, Decimal('36'))]

##### Compute the 7-day lag function on the sum column and save the result to lemon_sales_delay

In [20]:
%%sql
SELECT *, lag(sum, 7) OVER (ORDER BY sales_transaction_date) INTO
lemon_sales_delay FROM lemon_sales_sum;

 * postgresql://postgres:***@localhost/sqlda
2058 rows affected.


[]

In [22]:
%%sql 
SELECT * FROM lemon_sales_delay LIMIT 22;

 * postgresql://postgres:***@localhost/sqlda
22 rows affected.


[(datetime.date(2013, 5, 1), 6, Decimal('6'), None),
 (datetime.date(2013, 5, 2), 8, Decimal('14'), None),
 (datetime.date(2013, 5, 3), 4, Decimal('18'), None),
 (datetime.date(2013, 5, 4), 9, Decimal('27'), None),
 (datetime.date(2013, 5, 5), 9, Decimal('36'), None),
 (datetime.date(2013, 5, 6), 6, Decimal('42'), None),
 (datetime.date(2013, 5, 7), 8, Decimal('50'), None),
 (datetime.date(2013, 5, 8), 6, Decimal('56'), Decimal('6')),
 (datetime.date(2013, 5, 9), 6, Decimal('62'), Decimal('14')),
 (datetime.date(2013, 5, 10), 9, Decimal('71'), Decimal('18')),
 (datetime.date(2013, 5, 11), 3, Decimal('74'), Decimal('27')),
 (datetime.date(2013, 5, 12), 4, Decimal('78'), Decimal('36')),
 (datetime.date(2013, 5, 13), 7, Decimal('85'), Decimal('42')),
 (datetime.date(2013, 5, 14), 3, Decimal('88'), Decimal('50')),
 (datetime.date(2013, 5, 15), 3, Decimal('91'), Decimal('56')),
 (datetime.date(2013, 5, 16), 4, Decimal('95'), Decimal('62')),
 (datetime.date(2013, 5, 17), 6, Decimal('101'), Decimal('71')),
 (datetime.date(2013, 5, 18), 9, Decimal('110'), Decimal('74')),
 (datetime.date(2013, 5, 19), 6, Decimal('116'), Decimal('78')),
 (datetime.date(2013, 5, 20), 6, Decimal('122'), Decimal('85')),
 (datetime.date(2013, 5, 21), 11, Decimal('133'), Decimal('88')),
 (datetime.date(2013, 5, 22), 8, Decimal('141'), Decimal('91'))]

##### Calculate the growth rate using the data from lemon_sales_delay and store the resulting table in lemon_sales_growth. Label the growth rate column as volume

In [27]:
%%sql
SELECT *, ROUND(((sum-lag)/lag)*100) AS volume_percent INTO lemon_sales_growth FROM
lemon_sales_delay;

 * postgresql://postgres:***@localhost/sqlda
2058 rows affected.


[]

In [28]:
%%sql
SELECT * FROM lemon_sales_growth LIMIT 22;

 * postgresql://postgres:***@localhost/sqlda
22 rows affected.


[(datetime.date(2013, 5, 1), 6, Decimal('6'), None, None),
 (datetime.date(2013, 5, 2), 8, Decimal('14'), None, None),
 (datetime.date(2013, 5, 3), 4, Decimal('18'), None, None),
 (datetime.date(2013, 5, 4), 9, Decimal('27'), None, None),
 (datetime.date(2013, 5, 5), 9, Decimal('36'), None, None),
 (datetime.date(2013, 5, 6), 6, Decimal('42'), None, None),
 (datetime.date(2013, 5, 7), 8, Decimal('50'), None, None),
 (datetime.date(2013, 5, 8), 6, Decimal('56'), Decimal('6'), Decimal('833')),
 (datetime.date(2013, 5, 9), 6, Decimal('62'), Decimal('14'), Decimal('343')),
 (datetime.date(2013, 5, 10), 9, Decimal('71'), Decimal('18'), Decimal('294')),
 (datetime.date(2013, 5, 11), 3, Decimal('74'), Decimal('27'), Decimal('174')),
 (datetime.date(2013, 5, 12), 4, Decimal('78'), Decimal('36'), Decimal('117')),
 (datetime.date(2013, 5, 13), 7, Decimal('85'), Decimal('42'), Decimal('102')),
 (datetime.date(2013, 5, 14), 3, Decimal('88'), Decimal('50'), Decimal('76')),
 (datetime.date(2013, 5, 15), 3, Decimal('91'), Decimal('56'), Decimal('63')),
 (datetime.date(2013, 5, 16), 4, Decimal('95'), Decimal('62'), Decimal('53')),
 (datetime.date(2013, 5, 17), 6, Decimal('101'), Decimal('71'), Decimal('42')),
 (datetime.date(2013, 5, 18), 9, Decimal('110'), Decimal('74'), Decimal('49')),
 (datetime.date(2013, 5, 19), 6, Decimal('116'), Decimal('78'), Decimal('49')),
 (datetime.date(2013, 5, 20), 6, Decimal('122'), Decimal('85'), Decimal('44')),
 (datetime.date(2013, 5, 21), 11, Decimal('133'), Decimal('88'), Decimal('51')),
 (datetime.date(2013, 5, 22), 8, Decimal('141'), Decimal('91'), Decimal('55'))]

Sales growth of the Lemon Scooter

##### Observations
* We have calculated the cumulative sum, lag,
and relative sales growth of the Lemon Scooter. 
We can see that the initial sales volume is much larger than the other scooters, at over **800%**, and again finishes
higher at **55%**.
* The **Lemon Scooter**, over its production life cycle of 4.5 years, sold **16,558
units**, while the two Bat Scooters, the Original and Limited Edition models, sold **7,328**
and **5,803 units**, respectively, and are still currently in production, with the Bat Scooter
launching about 4 months earlier and with approximately 2.5 years of sales data
available.
* The **Original Bat Scooter**, which launched in October at a price of **599.99**,
experienced a **700%** sales growth in its second week of production and finished
the first 22 days with **28%** growth and a sales figure of 160 units.
* The **Bat Limited Edition Scooter**, which launched in February at a price of **699.99**,
experienced **450%** growth at the start of its second week of production and
finished with 96 sales and **66%** growth over the first 22 days.
* The 2013 Lemon Scooter, which launched in May at a price of **499.99**,
experienced **830%** growth in the second week of production and ended its first 22
days with 141 sales and **55%** growth.

##### Conclusions
Based on this information, we can make a number of different conclusions:

* The initial growth rate starting in the second week of sales correlates to the cost
of the scooter. As the cost increased to **699.99**, the initial growth rate dropped
from **830% to 450%.**
* The number of units sold in the first 22 days does not directly correlate to the
cost. The **599.99** Bat Scooter sold more than the 2013 Lemon Scooter in that first
period despite the price difference.
* There is some evidence to suggest that the reduction in sales can be attributed
to seasonal variations given the significant reduction in growth and the fact that
the original Bat Scooter is the only one released in October. So far, the evidence
suggests that the drop can be attributed to the difference in launch timing.